### ISSUE TITLE
Add a constraint for submultiplicity

### ISSUE THREAD

**Issue by monoidal** [tweag/ghc#426]
Currently, the solver approximates `p <= q` by `p ~ q`. Instead, we should emit a new type of constraint.

It could be simulated now by `sup p q = q`, but the inequality should be the primitive.

Out of scope for the initial version. Might be blocked on multiplicity casts. If we want to add a Zero multiplicity in the future, this will be necessary as constructors will be `One <= p => a # p -> Maybe a`.

With this constraint, we could generalize the type of constructors even if they are not linear. (Though I'm not sure if this is a good idea.)

**Comment by monoidal** [comment]
Moved to upstream:#19517.
